packages/astro/src/core/build/generate.ts

```diff
@@ -196,58 +198,35 @@ export async function generatePages(opts: StaticBuildOptions, internals: BuildIn
 		}
 	}
 
-	const staticImageList = getStaticImageList();
+	logger.info(null, dim(`Completed in ${getTimeStat(timer, performance.now())}.\n`));
 
-	if (staticImageList.size)
+	const staticImageList = getStaticImageList();
+	if (staticImageList.size) {
 		logger.info(null, `\n${bgGreen(black(` generating optimized images `))}`);
-	let count = 0;
-	for (const imageData of staticImageList.entries()) {
-		count++;
-		await generateImage(
-			pipeline,
-			imageData[1].options,
-			imageData[1].path,
-			count,
-			staticImageList.size
-		);
-	}
-
-	delete globalThis?.astroAsset?.addStaticImage;
 
-	await runHookBuildGenerated({
-		config: opts.settings.config,
-		logger: pipeline.getLogger(),
-	});
+		const totalCount = Array.from(staticImageList.values())
+			.map((x) => x.size)
+			.reduce((a, b) => a + b, 0);
+		const cpuCount = os.cpus().length;
+		const assetsCreationEnvironment = await prepareAssetsGenerationEnv(pipeline, totalCount);
+		const queue = new PQueue({ concurrency: cpuCount });
 
-	logger.info(null, dim(`Completed in ${getTimeStat(timer, performance.now())}.\n`));
-}
+		const assetsTimer = performance.now();
+		for (const [originalPath, transforms] of staticImageList) {
+			await generateImagesForPath(originalPath, transforms, assetsCreationEnvironment, queue);
+		}
 
-async function generateImage(
-	pipeline: BuildPipeline,
-	transform: ImageTransform,
-	path: string,
-	count: number,
-	totalCount: number
-) {
-	const logger = pipeline.getLogger();
-	let timeStart = performance.now();
-	const generationData = await generateImageInternal(pipeline, transform, path);
+		await queue.onIdle();
+		const assetsTimeEnd = performance.now();
+		logger.info(null, dim(`Completed in ${getTimeStat(assetsTimer, assetsTimeEnd)}.\n`));
 
-	if (!generationData) {
-		return;
+		delete globalThis?.astroAsset?.addStaticImage;
```

Comment:
```suggestion
		globalThis?.astroAsset?.addStaticImage = undefined;
```

Reply:
I remember you suggesting exactly this before (this is old code) and I don't remember why, but we couldn't do it for some reason... I'd have to check again

Reply:
Ah, probably TS saying that `addStaticImage` can't be `undefined`. That's fine, don't worry, it was just a suggestion :)

Here you can find the reasons: https://biomejs.dev/linter/rules/no-delete/